### ISSUE TITLE
When out of beta?

### ISSUE THREAD

**Issue by TheCrowned** [GoogleCloudPlatform/php-docs-samples#1723]
Universal Analytics is going to be retired in less than 2 months, and the PHP library is still in beta. When did Google turn into a sloppy startup? New documentation is also very unclear on the whole workflow. Tbh, I'm actively pushing my customers to move to Matomo or whatever other open platform - we're not enthusiastic of doing rushed implementation work in the next month (and even then, who knows when it will gets stable).

**Comment by bshaffer** [comment]
Hello there, are you referring to the [analytics data API](https://github.com/googleapis/php-analytics-data)? Could you provide a little more detail where you're encountering problems?

**Comment by TheCrowned** [comment]
Yes, I am referring to the Analytics API.

1. The API is in beta, and such has been for the last year. The base object is called `BetaAnalyticsDataClient` and all namespaces are prefixed with `Beta`. And yet Google has been actively pushing people towards GA4 for months. I get emails saying "hey, you should really use GA4, cause we will soon retire UA". I get customers asking when will we make our app compatible with GA4, and I keep having to point out that it's not even _stable_.
2. Where have authentication tokens gone? Their absence breaks our current flow. Our current setup is to have one single Cloud Project, keep the API credentials on our servers, and allow users to grant authorization to the app (which is a SaaS app). The result is that thousands of customers can use their own Analytics account and data by logging in once and getting a token from us, with a 20 seconds workflow. I can find mentions of `Tokens` in the [current docs](https://cloud.google.com/docs/authentication/token-types), but no examples anywhere, so I assume they are not supported anymore. If that's the case, now I have to tell customers to a) create a Cloud Platform project of their own; b) download the credentials json file; c) upload the file somewhere their app installation can access; d) add the user to their GA4 property to authorize it. This is just an _explosion_ in complexity, and I see no benefit at all for the final user, and implementation work on our side.
3. An astounding number of broken links. Take this page:  [Migrating to the Google Analytics Data API v1](https://developers.google.com/analytics/devguides/migration/api/reporting-ua-to-ga4) (mostly in `API schema changes` section). 
    - `list of dimensions and metrics` 404 https://developers.google.com/analytics/devguides/migration/api/api-schema
    - `custom dimension syntax` 404 https://developers.google.com/analytics/devguides/migration/api/api-schema#custom_dimensions
    - `custom metrics syntax` 404 https://developers.google.com/analytics/devguides/migration/api/api-schema#custom_metrics
    - `UA/GA4 API schema equivalence chart` 404 https://developers.google.com/analytics/devguides/migration/api/migration-api-schema-mapping
    - `numeric Google Analytics 4 property id` 404 https://developers.google.com/analytics/devguides/migration/api/property-id#what_is_my_property_id
    - ... probably more ...

This is a scenario I can imagine happening at a 100-people startup just graduating into a scale-up and struggling with major changes. But Google? Really? It's bewildering that Google doesn't even crawl its own docs to spot 404s. You deliver a **breaking change** that's gonna affect millions of websites, and 8 months from the time when you retire the Reporting API you don't even have the libraries out of beta, you require to largely re-visit current implementations, and the documentation is littered with 404s? 

(And I am even refraining from criticizing the choice of ditching the current API. You want to introduce a new API? Great, but leave the old one alone.)

**Comment by TheCrowned** [comment]
And on top of that, there seems to be some weird interference between GA4 properties and non-GA4 properties. I have some customers having both of them for the same account, and reporting that they get logged out of the app every few days. It did not use to happen before, and only happens for people who have some GA4 properties. Yet another malfunction.

**Comment by TheCrowned** [comment]
And it looks like there is no Adsense integration. A metric that used to be accessible in UA, and another breakage with the new GA4.

**Comment by meredithslota** [comment]
@TheCrowned Thank you for the detailed write-up (and finding all those broken links — 😭 that's embarrassing).

**Comment by meredithslota** [comment]
@bshaffer I'm not familiar with how those pages are authored — can you provide some guidance?

**Comment by AbdulazizSobh** [comment]
Why doesn't the Google team care to support and fix these problems
They are really going downhill and if they continue like this they will be a bad team
Please solve this problem related to this discussion because it is really important

**Comment by bshaffer** [comment]
Thank you @TheCrowned for reporting these 404s, I've added an internal bug (b/279435014) for this to be addressed (we maintain the PHP repositories, not the docs pages).

@meredithslota these are handled by the tech writers, I've filed a bug (see above). You can report a bug against any devsite page at any time using the toolbar in the bottom right

@AbdulazizSobh What support do you need here? This repo is for providing PHP samples for various google products. It seems the complaints in this bug refer to the Analytics Data product directly (or maybe I'm mistaken). So I am not sure there is anything else that needs to be done here.

You can see the full documentation for the PHP Analytics Data API here: https://cloud.google.com/php/docs/reference/analytics-data/latest

**Comment by TheCrowned** [comment]
> I can only say that this is not the place to ask these questions or log these complaints.

This is the best place I found, but please do direct me to a place where I can file my complaints. If anybody cares, that is.

**Comment by bshaffer** [comment]
@TheCrowned Check out the [Google Analytics Data API Support page](https://developers.google.com/analytics/devguides/reporting/data/v1/help), there's a link to a form where you can submit feedback. 

For other APIs, you can file issues on the [Cloud API issue tracker ](https://cloud.google.com/support/docs/issue-trackers) page. Hopefully we will add a link to an issue tracker for Google Analytics in the future.

**Comment by bshaffer** [comment]
@ikuleshov can also help answer these types of questions in the future;

**Comment by ikuleshov** [comment]
@TheCrowned 

I am reaching out on behalf of the Google Analytics Developer Relations team regarding your recent inquiry about the Google Analytics API. We would like to thank you for the feedback and will make sure such questions are routed to our team directly.

While there is currently no specific date for the GA4 APIs to become generally available, we do not expect breaking changes of existing functionality in the Beta versions of the Google Analytics Data API and Google Analytics Admin API. The primary reason why the APIs are still in the Beta channel is that new functionality is being added to the APIs as we speak.

Regarding your question about using authentication tokens in PHP. It should be possible to specify the tokens directly using the [credentials](https://cloud.google.com/php/docs/reference/analytics-data/latest/V1beta.BetaAnalyticsDataClient#:~:text=%E2%86%B3-,credentials,-string%7Carray%7CFetchAuthTokenInterface) parameter of BetaAnalyticsDataClient. It does not appear to be recommended to manipulate the tokens directly, though. For any PHP specific questions, kindly create an issue in the [Google APIs PHP repository](https://github.com/googleapis/google-cloud-php).

An update to allow the GA4 and AdSense linking is [currently in progress](https://support.google.com/adsense/answer/6084409).

As for the comment on GA vs GA4 interference, please feel free to contact the [Google Analytics Support Team](https://support.google.com/analytics/answer/9026876) for any general product related questions.

**Comment by TheCrowned** [comment]
Thanks @ikuleshov for replying, but I still don't see any concrete action in addressing any of the issues I pointed out. Not even the broken links in the docs have been fixed, crazy.

> While there is currently no specific date for the GA4 APIs to become generally available

What? UA is vanishing in less than 2 months, and you don't even have a _date_ for when the GA4 APIs would be stable? It should have been stable 1 year ago _already_.

> It should be possible to specify the tokens directly using the [credentials](https://cloud.google.com/php/docs/reference/analytics-data/latest/V1beta.BetaAnalyticsDataClient#:~:text=%E2%86%B3-,credentials,-string%7Carray%7CFetchAuthTokenInterface) parameter of BetaAnalyticsDataClient. It does not appear to be recommended to manipulate the tokens directly, though.

It _should_? It does not _appear_? Any developer expects clear docs with examples, not some vague info in an API reference. If you haven't been able to find it, so haven't I, and that's a problem. I'm not going to go through whatever repo source code to figure something out - I'm going to use (and recommend to customers) a competitor.
If you think _anybody_ is going to work with Beta APIs for their production environments ... I don't know where you live.

> An update to allow the GA4 and AdSense linking is [currently in progress](https://support.google.com/adsense/answer/6084409).

Again, what? There's not even an ETA specified! It should have been ready 1 year ago! It doesn't look like Google has any idea of the magnitude of the migration they are asking users to do.

> feel free to contact the [Google Analytics Support Team](https://support.google.com/analytics/answer/9026876) for any general product related questions.

I'm not gonna put in further efforts. There is plenty of information here if anybody wants to take action. If nobody wants, that's not my problem. The right course of action is to leave UA alone and build GA4 on the side, but it doesn't look like management likes this idea.
